File: app/src/main/java/com/adasa/foodplan/domain/model/MealPlanConfig.kt
package com.adasa.foodplan.domain.model

import kotlinx.datetime.DayOfWeek

data class MealPlanConfig(
    val schedule: MealScheduleConfig,
    val goals: NutritionGoals,
    val diet: DietPreferences,
    val rules: List<OptimizerRule>,
    val variety: VarietyConfig,
    val proteinPowder: ProteinPowder?,
    val shopping: ShoppingConfig
)

// ── Schedule ──────────────────────────────────────────────

data class MealScheduleConfig(
    val mealSlots: Map<DayOfWeek, DayMealConfig>,
    val batchGroups: List<BatchCookingGroup>,
    val snackOptionalFill: Boolean = true
)

data class DayMealConfig(
    val breakfast: Boolean = false,
    val lunch: Boolean = true,
    val dinner: Boolean = true,
    val snackCount: Int = 0
) {
    val isActive: Boolean get() = breakfast || lunch || dinner || snackCount != 0
}

data class BatchCookingGroup(
    val meal: MealCategory,
    val days: Set<DayOfWeek>,
    val batchNumber: Int        // 1, 2, 3 — matches color in the UI
)

// ── Nutrition goals ───────────────────────────────────────

data class NutritionGoals(
    val kcalTarget: Double,
    val proteinTarget: Double?,     // null = auto-calculated
    val fatTarget: Double?,
    val carbsTarget: Double?,
    val autoField: MacroField,      // which macro is derived from the others
    val minKcalPerDay: Double? = null,
    val maxKcalPerDay: Double? = null,
    val minProteinPerDay: Double? = null,
    val maxProteinPerDay: Double? = null,
    val minFatPerDay: Double? = null,
    val maxFatPerDay: Double? = null,
    val minCarbsPerDay: Double? = null,
    val maxCarbsPerDay: Double? = null
) {
    val resolvedProtein: Double get() = proteinTarget ?: ((kcalTarget - (fatTarget ?: 0.0) * 9 - (carbsTarget ?: 0.0) * 4) / 4).coerceAtLeast(0.0)
    val resolvedFat: Double get() = fatTarget ?: ((kcalTarget - (proteinTarget ?: 0.0) * 4 - (carbsTarget ?: 0.0) * 4) / 9).coerceAtLeast(0.0)
    val resolvedCarbs: Double get() = carbsTarget ?: ((kcalTarget - (proteinTarget ?: 0.0) * 4 - (fatTarget ?: 0.0) * 9) / 4).coerceAtLeast(0.0)
}

enum class MacroField { PROTEIN, FAT, CARBS }

// ── Diet preferences ──────────────────────────────────────

data class DietPreferences(
    val dietTypes: Set<DietType> = emptySet(),
    val allergies: Set<AllergyType> = emptySet(),
    val excludedIngredientIds: Set<String> = emptySet(),
    val preferredIngredientIds: Set<String> = emptySet(),
    val dislikedIngredientIds: Set<String> = emptySet()
)

enum class DietType(val displayName: String) {
    VEGETARIAN("Vegetarian"),
    VEGAN("Vegan"),
    PESCATARIAN("Pescatarian"),
    KETO("Keto")
}

enum class AllergyType(val displayName: String) {
    GLUTEN("Gluten"),
    DAIRY("Dairy"),
    NUTS("Nuts"),
    SHELLFISH("Shellfish"),
    EGGS("Eggs"),
    SOY("Soy"),
    PORK("Pork")
}

// ── Variety ───────────────────────────────────────────────

/**
 * Controls how aggressively the optimizer avoids repeating recipes.
 *
 * FLEXIBLE — small recency window, low penalty, no hard exclusions
 * BALANCED — moderate window, moderate penalty
 * STRICT   — large window, high penalty, recipes used this week are hard-excluded
 */
enum class VarietyLevel(val displayName: String, val description: String) {
    FLEXIBLE("Flexible", "Recipes can repeat often"),
    BALANCED("Balanced", "Moderate variety across weeks"),
    STRICT("Strict", "This week's recipes are completely blocked next week");

    val recencyWindowWeeks: Int get() = when (this) {
        FLEXIBLE -> 2
        BALANCED -> 4
        STRICT   -> 6
    }
    val recencyWindowDays: Int get() = recencyWindowWeeks * 7
    val penaltyWeight: Double get() = when (this) {
        FLEXIBLE -> 0.2
        BALANCED -> 0.6
        STRICT   -> 1.5
    }
    /** In STRICT mode, recipes within the window are hard-excluded from the pool. */
    val hardExcludeWithinWindow: Boolean get() = this == STRICT
}

/**
 * Per-category repeat limits. Both are batch-exempt: frozen slots are never
 * counted when enforcing these constraints.
 *
 * null = unlimited
 */
data class MealCategoryVariety(
    val maxTimesPerWeek: Int? = null,
    val maxConsecutiveDays: Int? = null
)

data class VarietyConfig(
    val level: VarietyLevel = VarietyLevel.BALANCED,
    val perCategory: Map<MealCategory, MealCategoryVariety> = defaultPerCategory(),
    val lunchDinnerSharedRecency: Boolean = true,
    val breakfastSnackSharedRecency: Boolean = false,
    val lunchDinnerMustDiffer: Boolean = true,
    val proteinSourceVariety: Boolean = true
) {
    /**
     * Derived from the two sharing toggles. Each set of categories shares one
     * recency index in the optimizer — a recipe used in any slot of the group
     * counts against all slots in that group.
     */
    val recencyGroups: List<Set<MealCategory>> get() = buildList {
        if (lunchDinnerSharedRecency) {
            add(setOf(MealCategory.LUNCH, MealCategory.DINNER))
        } else {
            add(setOf(MealCategory.LUNCH))
            add(setOf(MealCategory.DINNER))
        }
        if (breakfastSnackSharedRecency) {
            add(setOf(MealCategory.BREAKFAST, MealCategory.SNACK))
        } else {
            add(setOf(MealCategory.BREAKFAST))
            add(setOf(MealCategory.SNACK))
        }
    }
}

fun defaultPerCategory(): Map<MealCategory, MealCategoryVariety> = mapOf(
    MealCategory.BREAKFAST to MealCategoryVariety(maxTimesPerWeek = null, maxConsecutiveDays = null),
    MealCategory.LUNCH     to MealCategoryVariety(maxTimesPerWeek = 3,    maxConsecutiveDays = 2),
    MealCategory.DINNER    to MealCategoryVariety(maxTimesPerWeek = 2,    maxConsecutiveDays = 2),
    MealCategory.SNACK     to MealCategoryVariety(maxTimesPerWeek = 3,    maxConsecutiveDays = null)
)

// ── Protein powder ────────────────────────────────────────

data class ProteinPowder(
    val ingredientId: String,
    val name: String,
    val proteinPer100g: Double,
    val kcalPer100g: Double,
    val gramsInStock: Double,
    val autoFillGap: Boolean = true,
    val lowStockWarning: Boolean = true
) {
    val daysRemaining: Double get() = gramsInStock / 25.0
}

// ── Shopping ──────────────────────────────────────────────

data class ShoppingConfig(
    val shoppingDays: Set<DayOfWeek>,
    val intervalWeeks: Int = 1
)

File: flutter-app/lib/domain/model/meal_plan_config.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'recipe.dart';

part 'meal_plan_config.freezed.dart';

// ── Enums ─────────────────────────────────────────────────────────────────────

enum MacroField {
  protein,
  fat,
  carbs;

  String get firestoreName => name.toUpperCase();

  static MacroField fromFirestore(String? s) {
    if (s == null) return MacroField.protein;
    try {
      return MacroField.values.firstWhere(
        (e) => e.firestoreName == s.toUpperCase(),
      );
    } catch (_) {
      return MacroField.protein;
    }
  }
}

enum DietType {
  vegetarian,
  vegan,
  pescatarian,
  keto;

  String get firestoreName => name.toUpperCase();
}

enum AllergyType {
  gluten,
  dairy,
  nuts,
  shellfish,
  eggs,
  soy,
  pork;

  String get firestoreName => name.toUpperCase();
}

enum VarietyLevel {
  flexible,
  balanced,
  strict;

  int get recencyWindowWeeks => switch (this) {
    VarietyLevel.flexible => 2,
    VarietyLevel.balanced => 4,
    VarietyLevel.strict => 6,
  };

  int get recencyWindowDays => recencyWindowWeeks * 7;

  double get penaltyWeight => switch (this) {
    VarietyLevel.flexible => 0.2,
    VarietyLevel.balanced => 0.6,
    VarietyLevel.strict => 1.5,
  };

  bool get hardExcludeWithinWindow => this == VarietyLevel.strict;

  String get firestoreName => name.toUpperCase();
}

enum ConstraintType {
  minPerWeek,
  maxPerWeek;

  String get firestoreName =>
      name == 'minPerWeek' ? 'MIN_PER_WEEK' : 'MAX_PER_WEEK';

  static ConstraintType fromFirestore(String? s) {
    if (s == null) return ConstraintType.minPerWeek;
    return switch (s.toUpperCase()) {
      'MIN_PER_WEEK' => ConstraintType.minPerWeek,
      'MAX_PER_WEEK' => ConstraintType.maxPerWeek,
      _ => ConstraintType.minPerWeek,
    };
  }
}

enum RuleTargetType {
  dietCategory,
  ingredient;

  String get firestoreName =>
      name == 'dietCategory' ? 'DIET_CATEGORY' : 'INGREDIENT';

  static RuleTargetType fromFirestore(String? s) {
    if (s == null) return RuleTargetType.dietCategory;
    return switch (s.toUpperCase()) {
      'DIET_CATEGORY' => RuleTargetType.dietCategory,
      'INGREDIENT' => RuleTargetType.ingredient,
      _ => RuleTargetType.dietCategory,
    };
  }
}

enum AppTheme {
  system,
  light,
  dark;
}

// ── Domain models ─────────────────────────────────────────────────────────────

@freezed
class MealPlanConfig with _$MealPlanConfig {
  const factory MealPlanConfig({
    required MealScheduleConfig schedule,
    required NutritionGoals goals,
    required DietPreferences diet,
    required VarietyConfig variety,
    required List<OptimizerRule> rules,
    required ShoppingConfig shopping,
    ProteinPowder? proteinPowder,
  }) = _MealPlanConfig;

  factory MealPlanConfig.defaults() => MealPlanConfig(
        schedule: MealScheduleConfig.defaults(),
        goals: const NutritionGoals(kcalTarget: 1450.0, autoField: MacroField.protein),
        diet: const DietPreferences(),
        variety: const VarietyConfig(),
        rules: const [],
        shopping: ShoppingConfig.defaults(),
      );
}

// ── Schedule ──────────────────────────────────────────────────────────────────

@freezed
class MealScheduleConfig with _$MealScheduleConfig {
  const factory MealScheduleConfig({
    /// Key is ISO weekday: 1=Monday … 7=Sunday
    required Map<int, DaySlotConfig> perDay,
    @Default(<BatchCookingGroup>[]) List<BatchCookingGroup> batchCookingGroups,
    @Default(true) bool snackOptionalFill,
  }) = _MealScheduleConfig;

  factory MealScheduleConfig.defaults() => MealScheduleConfig(
        perDay: {
          for (int d = 1; d <= 7; d++)
            d: DaySlotConfig(
              hasBreakfast: false,
              hasLunch: true,
              hasDinner: true,
              snackCount: (d == 5 || d == 6) ? 1 : 0,
            ),
        },
      );
}

@freezed
class DaySlotConfig with _$DaySlotConfig {
  const factory DaySlotConfig({
    @Default(false) bool hasBreakfast,
    @Default(true) bool hasLunch,
    @Default(true) bool hasDinner,
    @Default(0) int snackCount,
  }) = _DaySlotConfig;
}

@freezed
class BatchCookingGroup with _$BatchCookingGroup {
  const factory BatchCookingGroup({
    required MealCategory category,
    /// ISO day numbers: 1=Monday … 7=Sunday
    required Set<int> days,
    required int batchNumber,
  }) = _BatchCookingGroup;
}

// ── Nutrition goals ───────────────────────────────────────────────────────────

@freezed
class NutritionGoals with _$NutritionGoals {
  const factory NutritionGoals({
    required double kcalTarget,
    double? proteinTarget,
    double? fatTarget,
    double? carbsTarget,
    @Default(MacroField.protein) MacroField autoField,
    double? minKcalPerDay,
    double? maxKcalPerDay,
    double? minProteinPerDay,
    double? maxProteinPerDay,
    double? minFatPerDay,
    double? maxFatPerDay,
    double? minCarbsPerDay,
    double? maxCarbsPerDay,
  }) = _NutritionGoals;

  const NutritionGoals._();

  double get resolvedProtein => proteinTarget ??
      ((kcalTarget - (fatTarget ?? 0.0) * 9 - (carbsTarget ?? 0.0) * 4) / 4)
          .clamp(0.0, double.infinity);

  double get resolvedFat => fatTarget ??
      ((kcalTarget - (proteinTarget ?? 0.0) * 4 - (carbsTarget ?? 0.0) * 4) / 9)
          .clamp(0.0, double.infinity);

  double get resolvedCarbs => carbsTarget ??
      ((kcalTarget - (proteinTarget ?? 0.0) * 4 - (fatTarget ?? 0.0) * 9) / 4)
          .clamp(0.0, double.infinity);
}

// ── Diet preferences ──────────────────────────────────────────────────────────

@freezed
class DietPreferences with _$DietPreferences {
  const factory DietPreferences({
    @Default(<DietType>{}) Set<DietType> dietTypes,
    @Default(<AllergyType>{}) Set<AllergyType> allergies,
    @Default(<String>{}) Set<String> excludedIngredientIds,
    @Default(<String>{}) Set<String> preferredIngredientIds,
    @Default(<String>{}) Set<String> dislikedIngredientIds,
  }) = _DietPreferences;
}

// ── Variety ───────────────────────────────────────────────────────────────────

@freezed
class MealCategoryVariety with _$MealCategoryVariety {
  const factory MealCategoryVariety({
    int? maxTimesPerWeek,
    int? maxConsecutiveDays,
  }) = _MealCategoryVariety;
}

@freezed
class VarietyConfig with _$VarietyConfig {
  const factory VarietyConfig({
    @Default(VarietyLevel.balanced) VarietyLevel level,
    Map<MealCategory, MealCategoryVariety>? perCategory,
    @Default(true) bool lunchDinnerSharedRecency,
    @Default(false) bool breakfastSnackSharedRecency,
    @Default(true) bool lunchDinnerMustDiffer,
    @Default(true) bool proteinSourceVariety,
  }) = _VarietyConfig;

  const VarietyConfig._();

  Map<MealCategory, MealCategoryVariety> get resolvedPerCategory =>
      perCategory ?? _defaultPerCategory;

  List<Set<MealCategory>> get recencyGroups => [
        if (lunchDinnerSharedRecency)
          {MealCategory.lunch, MealCategory.dinner}
        else ...[
          {MealCategory.lunch},
          {MealCategory.dinner},
        ],
        if (breakfastSnackSharedRecency)
          {MealCategory.breakfast, MealCategory.snack}
        else ...[
          {MealCategory.breakfast},
          {MealCategory.snack},
        ],
      ];
}

Map<MealCategory, MealCategoryVariety> get _defaultPerCategory => {
      MealCategory.breakfast: const MealCategoryVariety(),
      MealCategory.lunch:
          const MealCategoryVariety(maxTimesPerWeek: 3, maxConsecutiveDays: 2),
      MealCategory.dinner:
          const MealCategoryVariety(maxTimesPerWeek: 2, maxConsecutiveDays: 2),
      MealCategory.snack: const MealCategoryVariety(maxTimesPerWeek: 3),
    };

// ── Optimizer rules ───────────────────────────────────────────────────────────

@freezed
class OptimizerRule with _$OptimizerRule {
  const factory OptimizerRule({
    @Default('') String id,
    required RuleTargetType type,
    required String target,
    required String targetName,
    required ConstraintType constraint,
    required int value,
  }) = _OptimizerRule;
}

// ── Shopping config ───────────────────────────────────────────────────────────

@freezed
class ShoppingConfig with _$ShoppingConfig {
  const factory ShoppingConfig({
    required Set<int> shoppingDays,
    @Default(1) int intervalWeeks,
  }) = _ShoppingConfig;

  factory ShoppingConfig.defaults() =>
      const ShoppingConfig(shoppingDays: {7}); // Sunday = 7
}

// ── Protein powder ────────────────────────────────────────────────────────────

@freezed
class ProteinPowder with _$ProteinPowder {
  const factory ProteinPowder({
    required String ingredientId,
    required String name,
    required double proteinPer100g,
    required double kcalPer100g,
    required double gramsInStock,
    @Default(true) bool autoFillGap,
    @Default(true) bool lowStockWarning,
  }) = _ProteinPowder;

  const ProteinPowder._();

  double get daysRemaining => gramsInStock / 25.0;
}
